cou/steps/__init__.py

```diff
@@ -37,14 +37,27 @@ def __init__(
         self.params = params
         self.function = function
 
-    def add_step(self, step: UpgradeStep) -> None:
+    def add_step(self, step: UpgradeStep) -> UpgradeStep:
         """Add a single step."""
         self.sub_steps.append(step)
+        return self
 
-    def run(self) -> Any:
+    async def run(self) -> Any:
         """Run the function."""
         if self.function is not None:
             if self.params:
-                return self.function(**self.params)
-            return self.function()
+                return await self.function(**self.params)
+            return await self.function()
         return None
+
+    def __str__(self) -> str:
+        """Dump the plan for upgrade."""
+        result = ""
+        tab = "\t"
+        steps_to_visit = [(self, 0)]
+        while steps_to_visit:
+            step, indent = steps_to_visit.pop()
+            result += f"{tab * indent}{step.description}\n"
+            steps_to_visit.extend([(s, indent + 1) for s in step.sub_steps])
+
+        return result
```

Comment:
```suggestion
    def _to_string(self, ident: int=0) -> str:
        tab = "\t"
        self_string = f"{tab * ident}{self.description}"
        sub_steps_strings = [step._to_string(ident+1) for step in self.sub_steps]
        return os.linesep.join([self_string, *sub_steps_strings])

    def __str__(self):
        return self._to_string()
```

```python
>>> import os
>>> 
>>> class Step:
...     def __init__(self, desc, sub_steps = None):
...         self.description = desc
...         self.sub_steps = sub_steps or []
...     def _to_string(self, ident: int=0) -> str:
...         tab = "\t"
...         self_string = f"{tab * ident}{self.description}"
...         sub_steps_strings = [step._to_string(ident+1) for step in self.sub_steps]
...         return os.linesep.join([self_string, *sub_steps_strings])
...     def __str__(self):
...         return self._to_string()
... 
>>> test = Step("Parrent", [Step("S1"), Step("S2", [Step("B1")])])
>>> print(test)
Parrent
	S1
	S2
		B1
>>> print(test.sub_steps[0])
S1
>>> print(test.sub_steps[1])
S2
	B1
```

Reply:
I think that your version will not work if sub_step has another sub_step.

Reply:
It works :) I've added a new test case for you.

Reply:
No it's not, you just forgot to add `assert str(plan) == expected`.

```bash
=============================================== FAILURES ================================================
___________________________________ test___str__substep_has_substeps ____________________________________

    def test___str__substep_has_substeps():
        expected = (
            "Top level plan\n\tbackup mysql databases\n\t\tsub_step of sub_step\n\t\tanother "
            "sub_step of sub_step"
        )
        plan = UpgradeStep(description="Top level plan", parallel=False, function=None)
        substep = plan.add_step(
            UpgradeStep(description="backup mysql databases", parallel=False, function=None)
        )
        substep.add_step(
            UpgradeStep(description="sub_step of sub_step", parallel=False, function=None)
        )
        substep.add_step(
            UpgradeStep(description="another sub_step of sub_step", parallel=False, function=None)
        )
>       assert str(plan) == expected
E       AssertionError: assert 'Top level pl...l databases\n' == 'Top level pl...p of sub_step'
E           Top level plan
E         + 	another sub_step of sub_step
E         + 	sub_step of sub_step
E           	backup mysql databases
E         - 		sub_step of sub_step
E         - 		another sub_step of sub_step

tests/unit/steps/test_steps.py:92: AssertionError
----------------------------------------- Captured stdout call ------------------------------------------
Top level plan
	another sub_step of sub_step
	sub_step of sub_step
	backup mysql databases

```

Reply:
There was a a more series problem. steps.add was returning class instead of returning the added object. Both are fixed.

Reply:
Thanks :+1: